Invite notification fixes #959

MontrealSergiy opened the pull request:
This adds notification for invitation acceptance using elements of cbrain messaging system.

prioux commented on BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:
```diff
@@ -35,3 +35,16 @@
     </div>
   <%- end -%>
+
+  <div class="card">
```
Shouldn't the div be inside the `if` block ? Otherwise we get a div with nothing inside.

natacha-beck replied:
it should be inside the `unless` block.

natacha-beck replied:
+ same comment as @prioux made later:
> Change the unless something.blank? to if something.present? it looks much nicer and avoids double negatives.

prioux commented on BrainPortal/app/views/nh_invitations/_message_details.html.erb:
```diff
@@ -0,0 +1,52 @@
+   messclass = "#{prefix}_#{messtype}"
+%>
+
+<% unless mess.description.blank? %>
```
Change the `unless something.blank?` to `if something.present?` it looks much nicer and avoids double negatives.

MontrealSergiy replied:
ok Not sure is being blank something negative, `if` is shorter and as I know many devs consider `unless` bad style unless it removes a double negation.

natacha-beck commented on BrainPortal/app/controllers/invitations_controller.rb:
```diff
@@ -105,6 +105,14 @@ def update #:nodoc:
+    Message.send_message(@invitation.sender,
+                         :message_type   => 'notice',
+                         :header         => "Invitation Accepted",
+                         :description    => "A user joined project #{@nh_invitation.group.name}",
```
I confirm it make it crash.

natacha-beck commented on BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:
```diff
@@ -35,3 +35,16 @@
+    <%- for mess in @display_messages -%>
+      <p>
+      <div class="nh-invitations card-item">
+      <%= render(:partial  => "nh_invitations/message_details", :locals  => { :mess  => mess } ) rescue "Oh oh. your message was not rendered. Talk to devs!" %>
```
You do not need a rescue on a render. Rails will handle that.

In meantime we have the same code in `./messages/_message_display.html.erb` ==> ligne 33.

MontrealSergiy replied:
well, much code is similiar to similiar cbrain features. Do you say I should 'unrescue' old code too?

natacha-beck replied:
Can you give me example in CBRAIN ? I did not see them.

MontrealSergiy replied:
./messages/_message_display.html.erb ==> ligne 33.

natacha-beck replied:
Yep, I already mention it. Seems to be the 2 only place with this strange rescue. Please remove at both place.

MontrealSergiy replied:
besides the templates, much of message processing in neurohub_controller.rb is copy of cbrain_controller.

MontrealSergiy replied:
It does not handle well errors on my pc ![image](https://user-images.githubusercontent.com/11949081/90425669-1de8df00-e08e-11ea-83c3-b13676a81a95.png)

prioux replied:
Try in the Rails `production` environment

MontrealSergiy replied:
in production mode during malfunction the message seems silently hidden, which mean the user will do not know that his message got lost. Do you still want remove that rescues ? @natacha-beck

prioux replied:
Just use the same `rescue` that was there on the CBRAIN side, then.

prioux replied:
I don't want to waste too much time about this anymore.

MontrealSergiy replied:
As a compromise I removed the rescue form neurohub as users might understand that it is experimental project and confirmations could be lost

MontrealSergiy commented on BrainPortal/app/controllers/invitations_controller.rb:
```diff
@@ -105,6 +105,14 @@ def update #:nodoc:
     @invitation.save
 
     flash[:notice] = "You have been added to project #{@group.name}."
+
```
I guess @group will be best or at the least the shortest

MontrealSergiy commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -83,6 +84,36 @@ def check_if_locked
+    return unless current_user
+    return if     current_user.all_licenses_signed.blank?
+    return if     request.format.blank? || request.xhr?
+    return unless request.format.to_sym == :html || params[:controller] == 'messages'
```
well... I think putting && it will break the functionality (mind `unless`). As a compromise I can remove or comment out  the second part of the condition since, at the moment, there is no messages controller in NeuroHub part at the moment.

MontrealSergiy commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -83,6 +84,36 @@ def check_if_locked
+    @display_messages = []
+
+    unread_messages = current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
+    @unread_message_count = unread_messages.count
```
no reason. The possible future use is to put number of the unread messages over 'message' item of the navigation bar/top menu like Candice done. I guess can be deleted both from CBRAIN and NeuroHub controllers.

prioux replied:
Don't do that. Just keep the code clean. If someone in the future needs an instance variable, they'll add it then.

MontrealSergiy commented:
@natacha-beck actually you are right the rails handles templates error and produces an error message saying admins are notified of the issue. Regretfully for some reason when that happens neurohub redirects user to cbrains group page which seems strange.

![Screenshot 17 08 2020 22_47_19](https://user-images.githubusercontent.com/11949081/90466392-ae9bdb00-e0df-11ea-988f-34564d1a729d.png)

prioux commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -31,6 +31,7 @@ class NeurohubApplicationController < ApplicationController
 
   before_action :switch_to_neurohub_layout
   before_action :prepare_invites
+  before_action :prepare_messages
```
Now you're asking the rails framework to call `prepare_messages` twice for every request, because the callback is already registered in ApplicationController

prioux replied:
I checked and it's not called twice, Rails is intelligent enough not to invoke a method twice if you register it several time. But the point is still valid, it's already registered in ApplicationController.

prioux replied:
Please remove it

prioux commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -83,6 +84,35 @@ def check_if_locked
+  ########################################################################
+
+  # Find new messages to be displayed at the top of the page.
+  def prepare_messages #:nodoc:
```
This is a complete duplication of the method of the same name in Application controller, except for a single line. That is not DRY at all.

Please:

1. remove this method here since it's already defined in ApplicationController
2. change the method in ApplicationController such that the one line that differs invokes a method that can be overrided, which you will add to both files:
```ruby
   unread_messages = unread_messages_to_display
```

so that  in ApplicationController:

```ruby
# put this method just after prepare_messages
def unread_messages_to_display 
   current_user.messages.where( :read => false ).order( "last_sent DESC" )
end
```

and in NeurohubApplicationController we ONLY have this now:

```ruby
def unread_messages_to_display
  current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
end
```

MontrealSergiy replied:
added

prioux commented on BrainPortal/app/views/nh_invitations/_message_details.html.erb:
```diff
@@ -0,0 +1,52 @@
+
```
This file seems like aalso a almost complete duplication of `messages/_message_details.html.erb`, can we instead use THAT and not duplicate so much code?

MontrealSergiy replied:
Surely, it is easy to adjust that file. That what I done in an early version of the feature. I opted to allow Neurohub have it own template that can evolve independtly, but totally can drop that

prioux replied:
OK, I agree with this sentiment. Let's have a separate file for neurohub, like you did.

prioux replied:
You convinced me.

prioux commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -83,6 +84,35 @@ def check_if_locked
+
+    @display_messages = []
+
+    unread_messages = current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
```
This query is brittle by the way. If someone in the future change the header of the message that is being sent (see your code in InvitationControlelr and NhInvitationController), then this piece of code will stop displaying the message.

MontrealSergiy replied:
Should I then add `message_type=confirmation` flag along with relevant styles? 

Showing all the cbrain messages seems a bit risky. Or as a half measure may be add a unittest?

prioux replied:
Just use the same query but encspsulate it in a method, like I explained above, so that you only have one 'prepare_message' callback in ApplicationController.

prioux commented on BrainPortal/app/views/nh_invitations/_message_details.html.erb:
```diff
@@ -0,0 +1,52 @@
+%>
+
+<% if mess.description.present? %>
+  <p class=<%= messclass %>>
```
By the way, the values of the HTML attributes should be quoted. They were quoted in the original `messages/_message_details.html.erb`, why did you remove the quotes in your copy?

MontrealSergiy replied:
I tried a bit improve over existing code.

MontrealSergiy replied:
Note If I add quotes to this version I will get double quotes

prioux replied:
This is not fixed yet

MontrealSergiy replied:
added

prioux commented on BrainPortal/app/controllers/nh_invitations_controller.rb:
```diff
@@ -132,6 +132,13 @@ def update #:nodoc:
 
     flash[:notice] = "You have been added to project #{@nh_project.name}."
+
+    Message.send_message(@nh_invitation.sender,
```
By the way, here, why not use @nh_project.name instead of @nh_imvitation.group.name ?

MontrealSergiy replied:
ok that's easy. I experimented a bit to make cut and paste from cbrain which we often use for neurohub smoother ( you do not have to remember that groups became nh_projects, just add/remove nh). But can change.

prioux commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -122,5 +97,13 @@ def prepare_invites
+  # this patches prepare_message method
+  # to display 'Invitation Accepted' entitled messages. They will be show separately from Invitations
+  # and only kind of message to be shown for now.
+
```
You must not leave a blank line between the method description and the method, otherwise the documetnation code produces a method with no description

prioux replied:
I'd also suggest adding the #:nodoc: keyword because that method is not really a method that CBRAIN developer will want to invoke at other places.

prioux replied:
Same with the copy in ApplicationController

MontrealSergiy replied:
nodoc added

MontrealSergiy commented:
I think I addressed the remaining comments. I kept a query method description separated because, as you noticed, it is fragile but put the nodoc thing.

prioux commented on BrainPortal/app/controllers/neurohub_application_controller.rb:
```diff
@@ -91,6 +95,12 @@ def prepare_invites
   end
 
+  # below a method is overwriten to render only invitation confirmation, filtered by header (for now)
+  
```
Don't leave a blank line!
